Formspecs: Add state-selection to style elements

Df458 opened the pull request:
This PR addresses one of the points in #9044, extending formspec style properties by allowing the user to assign styles to specific element states. When a style has a state, it will only apply when the targeted element is in that state. For instance:
`style_type[button:hovered;textcolor=red]`
...will make text on a button turn red when the user hovers over it. States can also be combined like so:
`style_type[button:hovered,pressed;textcolor=red]`
This works like an AND, all listed states must be active for the style to apply.

For now, the following states are supported:

- `default` - Equivalent to no state
- `hovered` - Active when the user hovers over an element
- `pressed` - Active when an element is pressed

Additionally, states are only supported by buttons for now. I think this is fine for the time being, since no other elements support state-based style properties yet.

## Why add this?

Since we already have _hovered/_pressed variants for some style properties, this PR may not seem like a big deal. However, this is a much more flexible and scalable solution than continuing to add such properties, and will continue to create new possibilities as states and properties are added. There are a couple of interesting new features (such as changing text color based on state, and combining multiple states), and I'm already planning to add two additional states (active and disabled) in followup PRs. So, I think this is a very worthwhile PR.

## Technical notes

This PR works by tracking styles separately and combining them when they're needed for an element. Each `StyleSpec` has a state bitmask, and each element (that supports states) receives an array of styles, one for each state permutation. When elements change state, they re-apply the correct style to update their properties.

States take precedence over one another by numerical value. This means two things:
- Combination states take precedence over their individual parts
- The order of State enum values decides which states will take precedence over each other

I've tried to make the ordering as logical as possible: default < hovered < pressed < hovered & pressed

One notable weakness of the current approach is that as we add more states, the amount of work and memory involved in styling will dramatically increase. I don't expect us to add many states overall, I can't think of any good ones aside from the 4 that are planned/implemented so I don't think it's a big deal. Since this is a result of implementation details, we should be able to optimize further as it becomes necessary.

Theoretically, we could sneak this in for 5.2 and remove the _hovered/_pressed properties... but in the interest of not rushing things or screwing over folks who've started updating their mods, I've tried to maintain full compatibility with them instead. We can mark them as deprecated once this is in, and consider removal in a few releases (since unknown style properties don't cause errors).

## Testing

This is a pretty *deep* change: It affects anything using style properties, mainly buttons. I've tested the basics, ensuring that the test formspec is unchanged and making one or two styles in the new format, but ideally we need to test:
- Mods using state-based properties
- Formspecs using the extended `image_button` syntax with press feedback
- Ideally, every button property in every state. We should especially take care that property values reset correctly when a button's state changes back to default.

v-rob commented:
Wonderful!  But I can think of another future state: `checked` for checkboxes and potentially togglebuttons if we ever make those.

You said only buttons are supported.  Does this include `image_button` and `item_image_button` as well?

Df458 commented:
Good point on `checked`. And yes, all subclasses of `GUIButton` are supported!

Df458 commented:
From further testing, I've noticed one regression: Setting the `bgcolor` of a button for the default state propagates the value as-is, which means the element doesn't lighten/darken in the hovered/pressed states...

If we apply this concept to non-buttons, this behavior makes sense. However, it does force people to change their code. One possible alternative is to automatically lighten/darken the background according to a button's state regardless of what colors are passed, but this limits what users can do a little bit so I want to get a second opinion before implementing the change. Any thoughts?

rubenwardy commented:
I think if a hovered or pressed bgcolor isn't specified, it should make an adapted version from the default state. No flexibility it lost as elements can specify the color explicitly for hover and pressed

Df458 commented:
This could be done as well, but it will likely require notable structural changes to the PR to accomplish. Right now, the propagation occurs when the element is created (to avoid having to recalculate values every state change), but since this code is outside of the element it has no concept of "inherited" vs "specified" properties.

We'll either have to change StyleSpec to track this information, make a button-specific workaround in GUIFormspecMenu (absolutely not), or recalculate the style on state changes. The latter will probably have a notable perf impact, but also avoids the scalability problems so it might be the right call anyway.

Df458 commented:
I've made the change as described in my last comment. Testing seems to give the same results, but with the regression fixed. Let me know what you guys think.

v-rob commented:
So, you can combine selectors, but what about having multiple selectors, like #9380 does with the element types?  E.g.:
```
style_type[button:hovered+focus,pressed;textcolor=red]
```
`+` is what you currently have as `,`.  If this button is hovered and focused at the same time, or if it is pressed, the textcolor should be red.  This would be useful if you want all states of a button to appear the same, like this:
```
style_type[button:normal,hovered,pressed;bgcolor=red]
```
instead of the very long
```
style_type[button;bgcolor=red]
style_type[button:hovered;bgcolor=red]
style_type[button:pressed;bgcolor=red]
```
All buttons will have the same colors whether normal, hovered, or pressed.  This also would make another keyword like `normal` which is without any states at all and cannot be combined like `hovered+normal` since that makes no sense.

Df458 commented:
First of all, "normal" already exists: It's called `default`...
As for adding multiple selectors, my thought process is this: It adds nothing can't already be done, and increases the complexity of the code.
It would be a nice shortcut for modders, but it's a shortcut that could viably be added in a later PR to keep this one simpler. We're already dealing with some major changes, after all.

I'm not opposed to changing the specific syntax now to enable such an addition later. If we take an example from CSS, using `,` as an "or" instead of "and" certainly makes more sense so I'll change that symbol to something else.

v-rob commented:
Oh, didn't notice `default`.  I'm fine with just changing the syntax right now; I just want things to be designed with the future in mind.  It's not thinking about the future that has led to formspecs being so bad in the first place, so I want to prevent that as much as possible.

Df458 commented:
I've made the change, though while testing it a thought occurred to me:

What should we do if one state is unknown? Logging an error seems like a good idea, but should we throw out the whole style or just skip that state and move on?

rubenwardy commented:
> What should we do if a state is unknown? Logging an error seems like a good idea, but should we throw out the whole style or just skip that state and move on?

So, consider that an unknown state could be added if more states are supported in the future. In which case, it should ignore that definition as you don't want a style for a new state to affect old states. I'm unsure about logging, but definitely should not be an error - warning at most

Df458 commented:
Oh right, I didn't consider the backwards-compatibility angle... I think I'll log a warning and skip the definition. My thought re: logging was more from a validation standpoint, to make it immediately clear to modders if states are being correctly parsed. We don't ant to bother people with old clients though, so I think a warning would be the right call.

v-rob commented:
Food for thought: How are we going to handle scrollbar styling?  We have the thumb, up, left, right, and bottom arrows to handle.  Having no way to style them individually is sub-optimal, although I suppose they all should be styled together by default.

I think, if this is added via state selection, `:checked` could be reused for `:vertical`/`:horizontal` to select the type of scrollbar since scrollbars have no need of `:checked`.  We would need to distinguish between horizontal and vertical to tell what direction the arrows should be in.  Then, we would have to add selectors for the two different arrows.  This would add up to two new selectors, like `:arrows-top-right` and `:arrows-bottom-left`.  Possibly one also for the thumb.

So, I don't know how we would do scrollbars since 2/3 new selectors is rather hefty.

Df458 commented:
> Food for thought: How are we going to handle scrollbar styling?

This is in the future, but I see two ways:
- Child selectors/classes in the style of CSS (i.e. `style_type[scrollbar>button.button-left; ...]`)
- Control templates in the style of WPF (kinda hard to give an example, but this essentially comes down to users being able to redefine more complex elements using simpler ones)

Either way, state selectors are *absolutely not* the solution, because the attributes in question will never change.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -2496,15 +2496,25 @@ Elements
+* <state> is a comma-separated list of states.
+	* If a state that the element supports is provided, the style will only
+	  apply when the element is in that state.
+	* All provided states must be active for the style to apply
```
Please use spaces in the docs.

SmallJoker commented on src/gui/StyleSpec.h:
```diff
@@ -99,6 +109,49 @@ class StyleSpec
+
+	//! Using a list of styles mapped to state values, calculate the final
+	//  combined style for a state by propagating values in its component states
+	static StyleSpec getStyleFromStatePropagation(const std::array<StyleSpec, NUM_STATES>& styles, State state)
```
`const std::array<StyleSpec, NUM_STATES> &styles`

SmallJoker commented on src/gui/guiButton.cpp:
```diff
@@ -749,76 +744,99 @@ void GUIButton::setColor(video::SColor color)
+					Colors[i] = irr::video::SColor(Colors[i].getAlpha(),
+							core::clamp<u32>(Colors[i].getRed() * COLOR_HOVERED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getGreen() * COLOR_HOVERED_MOD, 0, 255),
+							core::clamp<u32>(Colors[i].getBlue() * COLOR_HOVERED_MOD, 0, 255));
```
This color modifying & clamping code should become a helper function.
I think either `numeric.cpp` or `StyleSpec.cpp` would be suited.

Df458 replied:
Good call, that was starting to get too repetitive.
I don't like putting this in StyleSpec because it's not closely-related enough (this is more default element behavior than styling behavior), though putting the function in numeric adds an extra include.

Not sure how I feel about that, but it's better than what was there before so I've done it.

SmallJoker commented on src/gui/guiButtonImage.cpp:
```diff
@@ -30,109 +30,48 @@ using namespace irr;
 
 	const core::position2di buttonCenter(AbsoluteRect.getCenter());
 	core::position2d<s32> geom(buttonCenter);
 	if (style.isNotDefault(StyleSpec::FGIMG)) {
```
Why is `FGIMG` not handled in `setFromStyle`, but `BGIMG` is?

Df458 replied:
I'm not sure I understand, both are handled. GUIButtonImage is the only element that supports `FGIMG`, so the code for handling it is here. The base class implementation handles the shared style properties, like `BGIMG`.

SmallJoker replied:
Oh, I missed such an obvious point. Thank you for explaining.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1894,35 +1886,42 @@ void GUIFormSpecMenu::parseImageButton(parserData* data, const std::string &elem
+		e->setStyles(style);
+		e->setScaleImage(true);
+
+		theme_by_name[spec.fname].erase(theme_by_name[spec.fname].end() - styles_to_pop, theme_by_name[spec.fname].end());
```
Line needs breaking (> 80 characters).

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2459,6 +2460,31 @@ bool GUIFormSpecMenu::parseStyle(parserData *data, const std::string &element, b
+				if (converted == StyleSpec::STATE_INVALID) {
+					infostream << "Unknown style state " << state <<
+							" in element '" << element << "'" << std::endl;
+					return false;
```
The `errorstream` log above doesn't return but this `infostream` does. I'd suggest to change this to `continue;`.

Df458 replied:
This should not be `continue`. The end-result of that would be breaking styling on old clients if new states were added, instead of correctly ignoring the states (For more details, see the short discussion with @rubenwardy earlier in the thread). You're right that the lack of return on the error is inconsistent though, I'll add one there.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2487,9 +2513,51 @@ bool GUIFormSpecMenu::parseStyle(parserData *data, const std::string &element, b
+	}
+
+	// Backwards-compatibility for existing _hovered/_pressed properties
+	if(spec.hasProperty(StyleSpec::BGCOLOR_HOVERED) || spec.hasProperty(StyleSpec::BGIMG_HOVERED) || spec.hasProperty(StyleSpec::FGIMG_HOVERED)) {
```
Please mark those enum entries as deprecated in `StyleSpec.h` so that it's clear they're only there for compatibility purposes.

Also: Missing space after `if`. Lines too long. Please split after `||` for a nice looking list.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4394,25 +4462,36 @@ std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
 		auto it = theme_by_type.find(parent_type);
 		if (it != theme_by_type.end()) {
-			ret |= it->second;
+			for(const StyleSpec& spec : it->second) {
```
`for (const StyleSpec &spec : it->second) {`

Also below.

SmallJoker commented on src/script/lua_api/l_mainmenu.cpp:
```diff
@@ -573,7 +573,8 @@ int ModApiMainMenu::l_show_keys_menu(lua_State *L)
 								-1,
-								engine->m_menumanager);
+								engine->m_menumanager,
+								engine->m_texture_source);
```
Would you mind changing this to 1 or 2 tabs for continued lines? This looks horrible.

SmallJoker commented:
Testing code:
1) minimal
2) [test_9slice.zip](https://github.com/minetest/minetest/files/4035237/test_9slice.zip) from #9290 (for compatibility tests)
<details>
<summary>3) Translated test_9slice to this PR</summary>

```Lua
local fs = [[
    size[9,4.5]
    real_coordinates[true]
    background9[0,0;2,4.5;fbg.png;true;3]

    style[tab;border=false;bgimg=bg3.png;bgimg_middle=3,2,29,29]
	style[tab:hovered;bgimg=bgh3.png]
	style[tab:pressed;bgimg=bgp3.png]
    button[3.5,0.75;1,0.5;tab;Button]
    button[4.5,0.75;1,0.5;tab;Button]
    image_button[3.5,1.75;1,1;fg.png;tab;]
    image_button[4.5,1.75;1,1;fg.png;tab;]
    item_image_button[3.5,3;1,1;default:apple;tab;]
    item_image_button[4.5,3;1,1;default:apple;tab;]

    style[3dscale;border=false;bgimg=bg.png]
	style[3dscale:hovered;bgimg=bgh.png]
	style[3dscale:pressed;bgimg=bgp.png]
    button[7,0.75;1,0.5;3dscale;Button]
    image_button[7,1.75;1,1;fg.png;3dscale;]
    item_image_button[7,3;1,1;default:apple;3dscale;]

    button[8.5,-0.5;1,1;3d;Button]
    button[8.5,1.75;1,1;default;Button]
]]

minetest.register_chatcommand("formspec", {
	func = function(name)
        minetest.show_formspec(name, "test", fs)
		return true
	end,
})
```
</details>

I compared these - no issues found.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -2496,15 +2496,25 @@ Elements
+### `style[<name>:<state>;<prop1>;<prop2>;...]`
 
 * Set the style for the named element `name`.
+* <state> is a comma-separated list of states.
```
`<state>` is interpreted as HTML. Please use \`state\` like above.
Also in below in `style_type`

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -2635,6 +2647,14 @@ Some types may inherit styles from parent types.
 
+### Valid States
+
+* <all elements>
```
Also interpreted as HTML. How about italics?
```
* *all elements*
```

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4394,25 +4467,36 @@ std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
+}
+
+std::array<StyleSpec, StyleSpec::NUM_STATES> GUIFormSpecMenu::getStyleForElement(const std::string &type,
+		const std::string &name, const std::string &parent_type) {
```
Please move the bracket to the next line.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4394,25 +4467,36 @@ std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
 		auto it = theme_by_type.find(parent_type);
 		if (it != theme_by_type.end()) {
-			ret |= it->second;
+			for(const StyleSpec &spec : it->second) {
```
missing space after `for`, superfluous brackets. Same 2x below.

Df458 commented:
All fixed, thanks for the review!

Df458 commented:
Rebased and tested, should work properly with multi-element selection now!
I also updated the API, I was still saying the state list was comma-separated but in fact it has been `+`-separated for a little while... that could've been bad.

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2495,16 +2495,26 @@ Elements
 
 * Set the style for the named element(s) `name`.
+* `state` is a list of states separated by the `+` character.
+    * If a state that the element supports is provided, the style will only
```
```suggestion
    * If a state is provided, the style will only take affect if the element
      is in that state. If an element doesn't support a state, it will never be
      in that state.
```

SmallJoker replied:
`If an element doesn't support a state, it will never be in that state.`
Ah yes, the floor is made out of floor.

The entire sentence could be removed for being obvious.

rubenwardy replied:
yeah, I don't know why the original text specified that

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2495,16 +2495,26 @@ Elements
 * Set the style for the named element(s) `name`.
+* `state` is a list of states separated by the `+` character.
+    * If a state that the element supports is provided, the style will only
+      apply when the element is in that state.
```
```suggestion
```

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2495,16 +2495,26 @@ Elements
+* `state` is a list of states separated by the `+` character.
+    * If a state that the element supports is provided, the style will only
+      apply when the element is in that state.
+    * All provided states must be active for the style to apply
```
```suggestion
    * All provided states must be active for the style to apply.
```

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2495,16 +2495,26 @@ Elements
           (default: infinite).
 
-### `style[<name 1>,<name 2>,...;<prop1>;<prop2>;...]`
+### `style[<name 1>,<name 2>;<prop1>;<prop2>;...]`
```
how about:

```
### `style[<selector1>,<selector2>,...;<prop1>;<prop2>;...]`

* A <selector> can be one of:
    * `<name>` - an element name
    * `<name>:<state>` an element name, a colon, and an element state
```

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2495,16 +2495,26 @@ Elements
+### `style[<name 1>:<state>,<name 2>:state;<prop1>;<prop2>;...]`
 
 * Set the style for the named element(s) `name`.
+* `state` is a list of states separated by the `+` character.
```
```suggestion
* `<state>` is a list of states separated by the `+` character.
```

Df458 replied:
Actually, @SmallJoker specifically asked me to change those to that format so that they better match the format of other formspec elements in the API.

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2547,8 +2557,10 @@ Styling Formspecs
+    style_type[<type 1>:<state>,<type 2>:<state>;<prop1>;<prop2>;...]
 
 Where a prop is:
 
```
maybe also make a note of state somewhere around here

rubenwardy commented:
few nitpicks in the docs, I'll look through the actual code sometime after lunch and before the heat death of the universe :)

Df458 commented:
Technical writing is not my forte, but I've made some adjustments that (hopefully) clear things up a little. Let me know if there's anything that still needs tweaking.

Df458 commented:
Rebased on 5.2

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1911,35 +1903,44 @@ void GUIFormSpecMenu::parseImageButton(parserData* data, const std::string &elem
-			video::ITexture *texture = m_tsrc->getTexture(image_name);
-			e->setForegroundImage(guiScalingImageButton(
-				Environment->getVideoDriver(), texture, geom.X, geom.Y));
+		// Push temporary styles with parameters
```
Rather than pushing temporary styles, I wonder if it would be better to copy the results of `getStyleForElement` and then push these values to it?

Df458 replied:
Yeah, good call. I think this wasn't possible when I first started the PR, and I just forgot to simplify things after that changed.

rubenwardy commented:
:+1: apart from that
